### PR TITLE
Feat: Enable explicit specification of SSO OAuth token_endpoint in manifest.yaml

### PR DESCRIPTION
**Pull request by jskelin**
Some customers have their own SSO OAuth systems connected with Dynatrace. Or like Dynatrace, have more than one OAuth authorization system. For that purpose, it must be possible to specify OAuth `tokenEndpoint` address. 

With this PR new section `tokenEndpoint` inside `oAuth` will be defined (see example). If `tokenEndpoint` is not provided, the default one `https://sso.dynatrace.com/sso/oauth2/token` will be used.

``` yaml
environmentGroups:
  - environments:
    - auth:
        oAuth:
          clientId:
            name: ENV_CLIENT_ID
          clientSecret:
            name: ENV_CLIENT_SECRET
          tokenEndpoint:
            value: "https://sso.dynatrace.com:443/sso/oauth2/token"
```

### CODE REVIEW

File: pkg/manifest/manifest_persitence.go [hunk 1]
```diff
@@ -39,8 +39,9 @@ type authSecret struct {
 }
 
 type oAuth struct {
-	ClientID     authSecret `yaml:"clientId"`
-	ClientSecret authSecret `yaml:"clientSecret"`
+	ClientID      authSecret `yaml:"clientId"`
+	ClientSecret  authSecret `yaml:"clientSecret"`
+	TokenEndpoint url        `yaml:"token_endpoint,omitempty"`
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest_persitence.go:44]: please make that `tokenEndpoint` , so we consistently use the same case in the YAML

Comment (jskelin) [reply pkg/manifest/manifest_persitence.go]: done

File: pkg/manifest/manifest_loader_test.go [hunk 2]
```diff
@@ -22,15 +22,18 @@ import (
...
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/spf13/afero"
+	"gopkg.in/yaml.v2"
 	"math"
 	"reflect"
 	"testing"
 
-	"gotest.tools/assert"
+	"github.com/stretchr/testify/assert"
+	assert2 "gotest.tools/assert"
```
Comment (jskelin) [inline pkg/manifest/manifest_loader_test.go:36]: Two complex cases left to migration on `github.com/stretchr/testify/assert`

``` go
assert2.DeepEqual(t, test.expectedManifest, mani, cmp.AllowUnexported(EnvironmentDefinition{}))
```

and 
``` go
assert2.DeepEqual(t, got, tt.want, cmpopts.SortSlices(func(a, b ProjectDefinition) bool { return a.Name < b.Name }))
```

**Comment by Laubi** [comment]
You have overlooked the code that we also need to write it in manifest_writer.go

File: pkg/manifest/manifest_loader_test.go
Comment (Laubi) [inline pkg/manifest/manifest_loader_test.go:1]: In this file, the test "TestLoadManifest" is the most important test, as it currently covers the actual loading with most code covered quite nicely.

Please add tests to that method that test if the tokenEndpoint.type is env, value, or invalid. Also a case where type=environment and the env-variable name can not be looked up or is empty

File: pkg/manifest/manifest_loader_test.go (hunk 2 above)
Comment (Laubi) [reply pkg/manifest/manifest_loader_test.go:36]: Since we agreed that we do either all or nothing in a file, I checked how we can do this, and it's really trivial to replace the code.
Simply use assert.Equal: `assert.Equal(t, got, tt.want)` and `assert.Equal(t, test.expectedManifest, mani)`

Please fix the code only to use testify

File: pkg/manifest/manifest_loader.go [hunk 3]
```diff
@@ -248,39 +221,65 @@ func parseOAuth(a oAuth) (OAuth, error) {
...
+		ClientId:      clientID,
+		ClientSecret:  clientSecret,
+		TokenEndpoint: urlDef,
 	}, nil
 }
 
-func parseManifestFile(context *ManifestLoaderContext, data []byte) (manifest, []error) {
-	var m manifest
-	var errs []error
+func getManifestYAML(context *ManifestLoaderContext) (manifest, error) {
```
Comment (Laubi) [inline pkg/manifest/manifest_loader.go:240]: The idea was to report all issues we can find at once to a user. This change now reports only one error, the user fixes it, and then the next error and so on. Why did you change it?

Also the name does not fit - I think parseManifest was quite good.

Comment (jskelin) [reply pkg/manifest/manifest_loader.go:240]: `getManifestYAML` just checks does file exists and is it a readable and if can, call unmarshall function

Logic from `parseManifestFile` is part of `verifyManifestYAML` which returns `[]error`. 

`getManifestYAML` is renamed to `readManifestYAML`

**Comment by jskelin** [comment]
> You have overlooked the code that we also need to write it in manifest_writer.go

Added writer and its tests

File: internal/oauth2/endpoints/endpoints.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package endpoints
+
+import (
+	"golang.org/x/oauth2"
+)
+
+var Dynatrace = oauth2.Endpoint{
```
Comment (UnseenWizzard) [inline internal/oauth2/endpoints/endpoints.go:23]: nitpick: I don't think there is any need for this to be an `oauth2.Endpoint`? 
If I'm not mistaken we only ever pull the TokenURL string out when loading and use it as string everywhere - so I think we could simply remove the unnecessary dependency on oauth2 and define a string in here.

File: pkg/manifest/manifest_loader.go [hunk 5]
```diff
@@ -248,39 +221,65 @@ func parseOAuth(a oAuth) (OAuth, error) {
 		return OAuth{}, fmt.Errorf("failed to parse ClientSecret: %w", err)
 	}
 
+	var urlDef UrlDefinition
+	if a.TokenEndpoint == nil {
+		urlDef = UrlDefinition{
+			Value: endpoints.Dynatrace.TokenURL,
```
Comment (UnseenWizzard) [inline pkg/manifest/manifest_loader.go:227]: I think here you need to set `Type: Default` for the logic in the writer to work

File: internal/oauth2/endpoints/endpoints.go (hunk 4 above)
Comment (UnseenWizzard) [reply internal/oauth2/endpoints/endpoints.go]: fine to keep, as it's standard practice in other libs/packages :+1:

File: pkg/manifest/manifest_loader.go (hunk 5 above)
Comment (jskelin) [reply pkg/manifest/manifest_loader.go:227]: added

**Review by UnseenWizzard — APPROVED** [review]
Looks good from my point of view. 

Please still run e2e tests as well. 

IMO we could merge this and change main if decide on Monday to move responsibility for the default URL to the `client` rather than `manifest` loading. 
Unless you think it's easier to change while still in the PR/never introducing the change to the manifest.
